rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,160 @@
+---
+authors: Marek Smoliński (marek@goteleport.com)
+state: draft
+---
+
+# RFD 33 - SNI and ALPN telepot proxy routing
+
+## What
+
+Combine all teleport proxy ports into one by routing proxy incoming traffic based on SNI and ALPN values.
+
+## Why
+
+Simplification of teleport proxy configuration, ability to expose ony one teleport proxy port 
+through firewall/load balancer.
+
+
+## Details
+
+| Port | Service | Description |
+| --- | --- | --- |
+| 3026  | **Proxy**  | HTTPS Kubernetes proxy proxy_service.kube_listen_addr  |
+| 3023  | **Proxy**  | SSH port clients connect to. A proxy will forward this connection to port #3022 on the destination node.  |
+| 3024  | **Proxy**  |  SSH port used to create "reverse SSH tunnels" from behind-firewall environments into a trusted proxy server. |
+| 3080  | **Proxy**  | HTTPS connection to authenticate tsh users and web users into the cluster. The same connection is used to serve a Web UI. |
+| MySQL Port | **Proxy**  | Handles MySQL client connection. |
+| Postgres Port| **Proxy**  | Handles Postgres client connection. |
+| 3022  | Node  | SSH port. This is Teleport's equivalent of port #22 for SSH.  |
+| 3025  | Auth  | SSH  port used by the Auth Service to serve its API to other nodes in a cluster.  |
+| 3025  | Kubernetes  | Kubernetes Service kubernetes_service.listen_addr  |
+
+In order to maintain backward compatibility a new proxy listener address (proxy_listener_addr) will be added to the proxy service alongside with current proxy listeners:
+to `teleport.yaml` configuration:
+```yaml
+proxy_service:
+   proxy_listener_addr: 0.0.0.0:443
```

Comment:
May be we can reuse the existing listen_addr and public_addr? We can add a new mode: field that will switch proxy to tls-sni mode:


```yaml
proxy_service:
   # new mode sets this proxy instance into TLS sni mode with listener using one port for all protocols
   mode: tls-sni
```

Reply:
This might work too, I was thinking about the scenario where users might want to use both "old-style" listeners and new "the one to rule them all" listener simultaneously - maybe during a migration/exploration phase before switching completely to the single TLS listener, cause technically nothing prevents them from being able to do it.

Reply:
Fair, but can they do that by deploying two proxy services with different configurations?

Reply:
In that case client or load balancer will be responsible for picking the correct proxy version.

Reply:
Yeah, I was also thinking about what Kevin mentioned as well - can we just make the new TLS listener backwards-compatible? So if SNI/ALPN is present, it will use the new dispatch logic and if not, fallback to current behavior. This way we won't need any new fields at all.